ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Codec.hs

```diff
@@ -21,7 +25,37 @@ import qualified Codec.CBOR.Read     as CBOR
 import           Text.Printf
 
 import           Ouroboros.Network.Codec
+import           Ouroboros.Network.Driver.Limits
 import           Ouroboros.Network.Protocol.TxSubmission.Type
+import           Ouroboros.Network.Protocol.Limits
+
+-- | Byte Limits.
+byteLimitsTxSubmission :: (bytes -> Word) -> ProtocolSizeLimits (TxSubmission txid tx) bytes
+byteLimitsTxSubmission= ProtocolSizeLimits stateToLimit
+  where
+    stateToLimit :: forall (pr :: PeerRole) (st :: TxSubmission txid tx).
+                    PeerHasAgency pr st -> Word
+    stateToLimit (ClientAgency (TokTxIds TokBlocking))    = largeByteLimit
+    stateToLimit (ClientAgency (TokTxIds TokNonBlocking)) = largeByteLimit
+    stateToLimit (ClientAgency TokTxs)                    = largeByteLimit
+    stateToLimit (ServerAgency TokIdle)                   = smallByteLimit
+
+
+-- | Time Limits.
+--
+-- `TokTxIds TokBlocking` No timeout
+-- `TokTxIds TokNonBlocking` `shortWait` timeout
+-- `TokTxs` `shortWait` timeout
+-- `TokIdle` `shortWait` timeout
+timeLimitsTxSubmission :: ProtocolTimeLimits (TxSubmission txid tx)
+timeLimitsTxSubmission = ProtocolTimeLimits stateToLimit
+  where
+    stateToLimit :: forall (pr :: PeerRole) (st :: TxSubmission txid tx).
+                    PeerHasAgency pr st -> Maybe DiffTime
+    stateToLimit (ClientAgency (TokTxIds TokBlocking))    = waitForever
+    stateToLimit (ClientAgency (TokTxIds TokNonBlocking)) = shortWait
+    stateToLimit (ClientAgency TokTxs)                    = shortWait
```

Comment:
I'm not sure why these are a short wait when similar chain sync ones are a long wait. 10s feels a bit short. I think I'd also go for long wait here, or 30s at least.

Reply:
With chainsync we may be waiting on a new block. For the short waits here we're waiting on receiving a short message.

Reply:
> With chainsync we may be waiting on a new block.

We're only waiting on a new block in the `StNext StMustReply` state, and that's why that state is `waitForever`. In all other states (except idle) we're "just" waiting on receiving a short message.

Reply:
Good point. I've updated chainsync to use shortWait instead of longWait.

Reply:
Ok. So the only long waits are in the block fetch, for blocks, but that's ok as they're big.